synse/websocket.go

```diff
@@ -0,0 +1,515 @@
+
+	resp := new(scheme.ResponseTags)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data.Tags, nil
+}
+
+// Info returns the full set of meta info and capabilities for a specific
+// device.
+func (c *websocketClient) Info(id string) (*scheme.Info, error) {
+	req := scheme.RequestInfo{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestInfo,
+		},
+		Data: scheme.WriteData{
+			Device: id,
+		},
+	}
+
+	resp := new(scheme.ResponseDevice)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// Read returns data from devices which match the set of provided tags
+// using ReadOptions.
+func (c *websocketClient) Read(opts scheme.ReadOptions) (*[]scheme.Read, error) {
+	req := scheme.RequestRead{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestRead,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseReading)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// ReadDevice returns data from a specific device.
+// It is the same as Read() where the label matches the device id tag
+// specified in ReadOptions.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) ReadDevice(id string, opts scheme.ReadOptions) (*[]scheme.Read, error) {
+	return nil, nil
+}
+
+// ReadCache returns stream reading data from the registered plugins.
+func (c *websocketClient) ReadCache(opts scheme.ReadCacheOptions) (*[]scheme.Read, error) {
+	req := scheme.RequestReadCache{
+		EventMeta: scheme.EventMeta{
+			ID:    addCounter(),
+			Event: requestReadCache,
+		},
+		Data: opts,
+	}
+
+	resp := new(scheme.ResponseReading)
+	err := c.makeRequest(req, resp)
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.verifyResponse(req.EventMeta, resp.EventMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resp.Data, nil
+}
+
+// WriteAsync writes data to a device, in an asynchronous manner.
+// NOTE - this method is not applicable for websocket client.
+func (c *websocketClient) WriteAsync(id string, opts []scheme.WriteData) (*[]scheme.Write, error) {
+	return nil, nil
+}
+
+// WriteSync writes data to a device, waiting for the write to complete.
+// FIXME - unlike http client's one, this does not support bulk write. Hence, there
+// should only be one WriteData value in the request opts. If multiple values are
+// specified, the first one would be picked. Same for the returned Transaction,
+// there should be only one value.
+// NOTE - this is related to compatibility issue mentioned in synse.go about the
+// need of uniforming client api.
```

Comment:
good catch. I'd open an issue on this. this could be an issue in the spec and we may need to amend it.